### ISSUE TITLE
conflicting cargo features?

### ISSUE THREAD

**Issue by evanrichter** [TheNeikos/crane-workspaces#9]
thoughts on being able to use crane-workspaces when some crates have conflicting features? e.g.:

```rust
#[cfg(feature = "a")]
pub type X = ();

#[cfg(feature = "b")]
pub type X = u8;
```

It's obviously not recommended use for features, but sometimes it's unavoidable. Currently I have [this branch](https://github.com/evanrichter/crane-workspaces/tree/no-feature-caching) that seems? to work for me now, but I think the solution is pretty fragile

I would love if there were

```nix
buildWorkspace {
  conflictingCargoFeatures = true;
  ...
}
```

that would disable the idealized caching that currently enables all features in dep crates

**Comment by TheNeikos** [comment]
I agree that the 'build with all dependencies' is not the correct choice for every workspace, and gets more and more wrong, the bigger the project.

The way I would like to see this solved is with multiple layers of feature sets that get built independently but in the same derivation.

So for example you could write:

```nix
let ws = cws.buildWorkspace {
  featureSets = {
    minimal = { 
      a = ["foo-feature" "bar-feature"]; 
      b = [ /* no features, this disables default features */]; 
    };
    withTime = { a = ["time-feature"]; };
  };
};
in
ws.workspaceMember.a.withFeatureSet "minimal"
```

Its a bit wordy right now, but the idea is that fundamentally you are meant to be saving compile cycles. So we have to share as much as possible. Since cargo automatically creates different files for different enabled features, we can 'share' a single derivation to build the deps of all featureSets. So in the worst case it is `O(nm)`, and in best case should be simply `O(n)`, with `n` the amount of deps, and `m` the list of feature sets.
